Add GCP VPC creation if no VPC available

infwinston opened the pull request:
<!-- Describe the changes in this PR -->

Fixes https://github.com/skypilot-org/skypilot/issues/1591

Problem: A fresh GCP project may not have any VPC created which is required for `sky launch`.
~~We add VPC creation to `sky check` if it's not found.~~

~~However, there's a design choice to make. What's the VPC name we should use, if there are many?~~
Currently, Ray just blindly uses the first VPC net in the order, which is obviously a bad idea. what if it's a user created one without sufficient permissions?
https://github.com/skypilot-org/skypilot/blob/3cceb5cf48943782ddfb7683ccad72a8f42d0f6f/sky/skylet/providers/gcp/config.py#L509-L517

We can use `default`, but the risk is the VPC with such name may not have sufficient permissions for skypilot to work.
Now, I try to create a dedicated VPC net with a unique name `skypilot-vpc` if such VPC is not found when running `ray up`.
~~However, the consequence is that after this PR the GCP users need to re-run `sky check` to create the VPC.~~


<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):
- [x] Tested `sky launch` with a fresh GCP project without any VPC
- [x] Tested a project with VPC + sufficient permissions
- [x] Tested a project with VPC + insufficient permissions
- [x] Tested TPU Node
- [x] `pytest tests/test_smoke.py --gcp` passed

concretevitamin reviewed (commented):
Q: is there a reason we put these two commands in gcp.py / `sky check` code path, rather than putting them in config.py? E.g., detect whether the VPC exists, if not, create them. That seems like a better place to handle these stack setup issues.

infwinston commented:
> Q: is there a reason we put these two commands in gcp.py / `sky check` code path, rather than putting them in config.py? E.g., detect whether the VPC exists, if not, create them. That seems like a better place to handle these stack setup issues.

Good point. I've moved the code that creates a VPC net + firewall rules to `ray up`. Also I switched from `gcloud` to restful API implementation, which is better I think.

infwinston commented:
OK I just realize a possible issue with this change. This will add another IAM permission requirement `compute.networks.create` for creating a VPC network. which is not needed before as Ray just blindly choose the first VPC net as I mentioned.
https://github.com/skypilot-org/skypilot/blob/3cceb5cf48943782ddfb7683ccad72a8f42d0f6f/sky/skylet/providers/gcp/config.py#L509-L517

I'm not entirely sure if this additional permission would create barriers to new users. but in some cases (e.g., no vpc or vpc without sufficient firewall rules) this permission is essential.
One idea is that we can still merge this and in IAM doc we ask those users without such permission to ask their admins to create a VPC net `skypilot-vpc` with essential firewall rules.

Michaelvll commented:
Great findings! Is it possible to do it with best effort, i.e try to create the network and error out hinting the user either adding the permission or manually do it?

Michaelvll commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -506,10 +507,50 @@ def _configure_subnet(config, compute):
 
     if not subnets:
-        raise NotImplementedError("Should be able to create subnet.")
+        # Create a default VPC network and firewall rules
```
```suggestion
        # SkyPilot: Create a default VPC network and firewall rules
```

infwinston replied:
Added before we call `get_usable_vpc(config)`

Michaelvll commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -506,10 +507,50 @@ def _configure_subnet(config, compute):
         return config
 
-    subnets = _list_subnets(config, compute)
+    subnets = _list_subnets(config, compute, filter=f'(name="{SKYPILOT_VPC_NAME}")')
```
Instead of always using the `SKYPILOT_VPC_NAME`, should we use the original way as before and use this creation as a backup:
1. Check all subnets and use the first possible subnet if any subnet exists
2. If no subnets exist, try to create a `SKYPILOT_VPC_NAME`

Reason:
1. Backward compatibility for any existing user, as a cluster launched with the default VPC will be stopped and relaunched if we use a different subnet, IIUC.
2. To avoid asking all the users to have permission to create a subnet, i.e. the proposal above will only require the user who does not have any subnets available to have that permission.

Michaelvll commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -506,10 +507,50 @@ def _configure_subnet(config, compute):
+        vpc_network_body = {
+            "name": SKYPILOT_VPC_NAME,
+            "selfLink": f"projects/{proj_id}/global/networks/{SKYPILOT_VPC_NAME}",
+            "autoCreateSubnetworks": True,
+            "mtu": 1460,
+            "routingConfig": {"routingMode": "GLOBAL"},
```
For future maintenance, can we add some comments for these arguments, such as where we get these argument value?

infwinston replied:
Good point! These are default parameters from GCP. Added comments to `constants.py`
```
# Below parameters are from the default VPC on GCP.
# https://cloud.google.com/vpc/docs/firewalls#more_rules_default_vpc
```

Michaelvll reviewed (commented):
Thanks for adding the VPC creation support @infwinston! Left serval comments below. The main concern for the current implementation are:
1. the new permission requirement for the users who already have their own default subnet with a different name as `SKYPILOT_VPC_NAME`.
2. the backward compatibility for the GCP cluster already launcehd.

infwinston commented:
Thanks @Michaelvll ! Those are great points. I've implemented a best effort search for VPC available. 
We first check if there's a VPC available with sufficient permissions. If so, we'll directly use it. If not, we'll create one with name `SKYPILOT_VPC_NAME`. so it should be backward compatible now.
It's a bit tricky when doing firewall rules comparisons, but I think current implementation should work. I also confirm that with our `skycamp-gcp` project, it'll directly use the `default` VPC rather than creating a new one.

Please take a look, thanks :)

infwinston commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,73 @@ def _configure_key_pair(config, compute):
+
+    def _get_refined_rule(rule):
+        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
+        return {k: rule[k] for k in KEY_TO_COMPARE}
```
Ah yes.. I did try the exactly same code. but the problem is `rule[k]` can be a list of dictionary in which is not hashable. so python would raise errors..
```
>>> dc = [{}, {}]
>>> set(dc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

I did check the order of protocols returned by GCP. It seems always to be a fixed order (tcp, ucp, then icmp}. But we may do this to guarantee the order. wdyt?
```
    def _get_refined_rule(rule):
        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
        refined_rule = {}
        for k in KEY_TO_COMPARE:
            if k == "allowed":
                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
            else:
                refined_rule[k] = rule[k]
        return refined_rule
```

Michaelvll replied:
The new implementation looks good to me.

infwinston commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,73 @@ def _configure_key_pair(config, compute):
+        return {k: rule[k] for k in KEY_TO_COMPARE}
+
+    required_rules = list(map(_get_refined_rule, required_rules))
+    effective_rules = list(map(_get_refined_rule, effective_rules))
```
ah good point!

infwinston replied:
ah okay it's the same problem. I think we just have to use `list`
```
    effective_rules = set(map(_get_refined_rule, effective_rules))
TypeError: unhashable type: 'dict'
```

Michaelvll replied:
That makes sense. Keep the original version looks good to me.

infwinston commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -506,10 +573,8 @@ def _configure_subnet(config, compute):
-
-    if not subnets:
-        raise NotImplementedError("Should be able to create subnet.")
+    usable_vpc_name = get_usable_vpc(config)
```
Ah yes, in `get_usable_vpc() function, there's a `_create_vpcnet()` which requires the VPC creation permission. Perhaps the function name is a bit misleading?
also, we can try to `try except` here to catch errors and give better messages to hint users that they need to contact admin.

Michaelvll replied:
Oh, I mean whether the `_list_vpcnets` function requires more permission.

Raising our own error when the permission is not sufficient makes sense to me, but as mentioned above, we need to be careful about the problem in https://github.com/skypilot-org/skypilot/pull/1594/files#r1073976728

infwinston replied:
got it. my best guess is `compute.subnetworks.list` (which is already used before this PR) and `compute.networks.list` (used to list a VPC net) should come together. so it shouldn't require more permission I think.

infwinston commented:
@Michaelvll OK. when I was testing the required permissions, I realized I was pretty much wrong..

The `Compute Admin` role actually contains the permission `compute.networks.create` for creating a new VPC net...
so in fact we don't really add more requirements to our assumption. 
ref: https://cloud.google.com/compute/docs/access/iam#compute.admin

Sorry for the mistake. but nonetheless, this PR is still needed and should work.

infwinston commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,73 @@ def _configure_key_pair(config, compute):
+            usable_vpc_name = vpc["name"]
+            break
+
+    if usable_vpc_name is None:
+        logger.info(f"Creating a default VPC network, {SKYPILOT_VPC_NAME}...")
+
+        # Create a default VPC network
+        proj_id = config["provider"]["project_id"]
+        body = VPC_TEMPLATE.copy()
+        body["name"] = body["name"].format(VPC_NAME=SKYPILOT_VPC_NAME)
+        body["selfLink"] = body["selfLink"].format(PROJ_ID=proj_id, VPC_NAME=SKYPILOT_VPC_NAME)
+        _create_vpcnet(config, compute, body)
+
+        # Create firewall rules
+        for rule in FIREWALL_RULES_TEMPLATE:
+            body = rule.copy()
+            body["name"] = body["name"].format(VPC_NAME=SKYPILOT_VPC_NAME)
+            body["network"] = body["network"].format(PROJ_ID=proj_id, VPC_NAME=SKYPILOT_VPC_NAME)
+            body["selfLink"] = body["selfLink"].format(PROJ_ID=proj_id, VPC_NAME=SKYPILOT_VPC_NAME)
+            _create_firewall_rule(config, compute, body)
+
+        usable_vpc_name = SKYPILOT_VPC_NAME
+        logger.info(f"A VPC network {SKYPILOT_VPC_NAME} created.")
```
As I replied in the main thread, no additional permission is needed for creating a VPC net. So I think this should be fine?

Michaelvll replied:
Sounds good. We can rely on the user respecting the doc for now. The problem still exists if the permission is not enough, and the cluster will never be able to be removed from the cluster table as the similar problem in #1575 .

infwinston replied:
Ah I see. I tested with creating a user without the `Compute Admin` role. There will be lots of permission errors raised and the cluster state will be in `INIT`. I then couldn't really `sky down` it as you mentioned.
I think this is a bigger problem maybe let's solve it in another PR?

Michaelvll replied:
Sounds good. Let's file an issue for that and fix it in the future.

Michaelvll commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,73 @@ def _configure_key_pair(config, compute):
+
+    usable_vpc_name = None
+    for vpc in vpcnets_all:
+        if _check_firewall_rules(vpc["name"], config, compute):
```
Note (may not need additional action): The user's previous clusters may be launched with another VPC, which doesn't have the exact same firewall rules, e.g. without some specific ports open. Adding this check here may still cause a resource leakage in that case. It might be fine for now, as we assume our existing users are using the default VPC created by GCP.

infwinston replied:
Ah you're right. But hopefully this is fine as existing users are just using default one.

infwinston commented on sky/backends/backend_utils.py:
```diff
@@ -44,6 +44,7 @@
 from sky.backends import onprem_utils
 from sky.skylet import constants
 from sky.skylet import log_lib
+from sky.skylet.providers.gcp import config as gcp_config
```
Fixed!

infwinston commented on sky/skylet/providers/gcp/constants.py:
```diff
@@ -0,0 +1,65 @@
+        }],
+        "sourceRanges": ["0.0.0.0/0"],
+    }
+]
```
Fixed!

infwinston commented:
@Michaelvll thanks again for the reviews. I've tested this PR with 
```
pytest tests/test_smoke.py --gcp
```
and fixed a bug! (when launching a TPU VM, a weird firewall rule will be added to the VPC..)
also thanks @Michaelvll for the smoke tests improvement.

Now the GCP smoke tests are all passed. Please let me know if it's good to go.

infwinston commented:
@Michaelvll @concretevitamin I've added the code to ensure there's no resource leakage. Basically we'll reuse the VPC if the VM is already launched.
(by listing all the instances and match them by cluster name)
https://github.com/skypilot-org/skypilot/blob/b4cfd2be60929be3a7ab1ae6d32fdde4ed15ae7c/sky/skylet/providers/gcp/config.py#L535-L544

I've also tested with below commands and confirmed no resource leaked after this change.
```
git checkout master
sky launch -c tmp --cloud gcp
git checkout fix-gcp-vpc
sky launch -c tmp --cloud gcp
```
Please let me know if there's any additional thoughts.

Michaelvll commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,93 @@ def _configure_key_pair(config, compute):
+    _, _, compute, _ = construct_clients_from_provider_config(config["provider"])
+
+    # For backward compatibility, reuse the VPC if the VM is launched.
+    ins_all = _list_instances(config, compute)
+    for ins in ins_all:
+        labels = ins.get("labels", {})
+        for k, v in labels.items():
+            if k == "ray-cluster-name" and v == config["cluster_name"]:
+                if "networkInterfaces" not in ins or len(ins["networkInterfaces"]) == 0:
+                    continue
+                vpc_name = ins["networkInterfaces"][0]["network"].split("/")[-1]
+                return vpc_name
```
Just curious, why don't we directly use the `node._list_instances()`?

infwinston replied:
Ah yes.. I tried to use it but found that the abstraction there isn't really what we want.
The semantic of `node._list_instances()` is not listing all VMs in the zone but they put a cluster name filter
https://github.com/skypilot-org/skypilot/blob/14e26d0bd8faf9e62008f195de93b94392f8fbb8/sky/skylet/providers/gcp/node.py#L409-L411
so it will only list the VM with the associated cluster name..

I guess it's simpler that we just implement it ourselves.

Michaelvll replied:
Sorry, I am confused. We also tries to filter out the clusters with a label that contains cluster_name, right?

infwinston replied:
ah I just updated the code that uses `_list_instances()`! I misunderstood some details of the function.

Michaelvll reviewed (approved):
Thanks for adding this @infwinston! It looks good to me now. This should unblock many of our new users. Cc'ing @concretevitamin, @romilbhardwaj in case they want to double-check it.

infwinston commented on sky/skylet/providers/gcp/config.py:
```diff
@@ -488,6 +488,96 @@ def _configure_key_pair(config, compute):
+    vpcnets_all = _list_vpcnets(config, compute)
+
+    usable_vpc_name = None
+    for vpc in vpcnets_all:
```
I feel like it'd be safer to still check the firewall rules? in case when users accidentally touch the rules set for `skypilot-vpc` and cause a silent bug.

Michaelvll replied:
Sounds good. If the user can accidentally modify the `skypilot-vpc`, is the operation below at L557-L576 idempotent, i.e. will it cause error if the `skypilot-vpc` already exists but does not have the correct firewall rules?

infwinston replied:
That's a very good point. Right now if user modify `skypilot-vpc`, we'll error out. But I guess this might still be better than a silent error?
I investigated a bit and had offline discussion with @Michaelvll. The solution to make it idempotent adds some additional complexity (validating/editing firewall rules), I think we should do it. But given this PR is urgent and blocking new GCP users, we think merging it first and make another PR for idempotence would be better.
